drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,747 @@
+		err = tmc5041_write(config->controller, TMC5041_RAMPMODE(config->index),
+				    TMC5041_RAMPMODE_POSITIVE_VELOCITY_MODE);
+		if (err != 0) {
+			return -EIO;
+		}
+		err = tmc5041_write(config->controller, TMC5041_VMAX(config->index), velocity_fclk);
+		if (err != 0) {
+			return -EIO;
+		}
+		break;
+
+	case STEPPER_DIRECTION_NEGATIVE:
+		err = tmc5041_write(config->controller, TMC5041_RAMPMODE(config->index),
+				    TMC5041_RAMPMODE_NEGATIVE_VELOCITY_MODE);
+		if (err != 0) {
+			return -EIO;
+		}
+		err = tmc5041_write(config->controller, TMC5041_VMAX(config->index), velocity_fclk);
+		if (err != 0) {
+			return -EIO;
+		}
+		break;
+	}
+
+	if (config->sg_threshold_velocity) {
+		k_work_reschedule(&data->stallguard_dwork,
+				  K_MSEC(config->sg_activation_delay_in_ms));
+	}
+	return 0;
+}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+
+int tmc5041_stepper_set_ramp(const struct device *dev,
+			     const struct tmc_ramp_generator_data *ramp_data)
+{
+	LOG_DBG("Stepper motor controller %s set ramp", dev->name);
+	const struct tmc5041_stepper_config *config = dev->config;
+	int err;
+
+	err = tmc5041_write(config->controller, TMC5041_VSTART(config->index), ramp_data->vstart);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_A1(config->index), ramp_data->a1);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_AMAX(config->index), ramp_data->amax);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_D1(config->index), ramp_data->d1);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_DMAX(config->index), ramp_data->dmax);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_V1(config->index), ramp_data->v1);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_VMAX(config->index), ramp_data->vmax);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_VSTOP(config->index), ramp_data->vstop);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_TZEROWAIT(config->index),
+			    ramp_data->tzerowait);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_VHIGH(config->index), ramp_data->vhigh);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_VCOOLTHRS(config->index),
+			    ramp_data->vcoolthrs);
+	if (err != 0) {
+		return -EIO;
+	}
+	err = tmc5041_write(config->controller, TMC5041_IHOLD_IRUN(config->index),
+			    ramp_data->iholdrun);
+	if (err != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#endif
+
+static int tmc5041_init(const struct device *dev)
+{
+	LOG_DBG("TMC5041 stepper motor controller %s initialized", dev->name);
+	struct tmc5041_data *data = dev->data;
+	const struct tmc5041_config *config = dev->config;
+	int err;
+
+	k_sem_init(&data->sem, 0, 1);
```

Comment:
You could add a static declaration of the semaphore to the device definition macro and only store a pointer to it in the data, this would remove the need for this initialisation. Or just combine the two lines by setting the initial count to 1 straight away.